mafContrib/kneMetrics.py

```diff
@@ -77,74 +84,226 @@ def __init__(self, metricName='KNePopMetric', mjdCol='observationStartMJD', m5Co
+        """
+        Detected at least 4 times in either izy colors
+
+        Parameters
+        ----------
+        mags : array
+            magnitudes obtained interpolating models on the dataSlice
+        t : array
+            relative times
+        min_det : float or int
+            minimum number of detections required in izy bands
+        """
+        result = 1
+        # Detected
+        around_peak = np.where((t > 0) & (t < 30) & (mags < dataSlice[self.m5Col]))[0]
+        # Filters of detected points
+        filters = dataSlice[self.filterCol][around_peak]
+        # Number of detected points in izy bands
+        n_red_det = np.size(np.where(filters == 'i')[0]) + np.size(np.where(filters == 'z')[0]) + np.size(np.where(filters == 'y')[0])
+        # Condition
+        if n_red_det < min_det:
+            return 0
+
+        return result
+
+    def _blue_color_detect(self, dataSlice, mags, t, min_det=4):
+        """
+        Detected at least 4 times in either ugr colors
+
+        Parameters
+        ----------
+        mags : array
+            magnitudes obtained interpolating models on the dataSlice
+        t : array
+            relative times
+        min_det : float or int
+            minimum number of detections required in ugr bands
+        """
+        result = 1
+        # Detected
+        around_peak = np.where((t > 0) & (t < 30) & (mags < dataSlice[self.m5Col]))[0]
+        # Filters of detected points
+        filters = dataSlice[self.filterCol][around_peak]
+        # Number of detected points in ugr bands
+        n_blue_det = np.size(np.where(filters == 'u')[0]) + np.size(np.where(filters == 'g')[0]) + np.size(np.where(filters == 'r')[0])
+        # Condition
+        if n_blue_det < min_det:
+            return 0
+
+        return result
+
     def run(self, dataSlice, slicePoint=None):
         result = {}
         t = dataSlice[self.mjdCol] - self.mjd0 - slicePoint['peak_time']
         mags = np.zeros(t.size, dtype=float)
 
         for filtername in np.unique(dataSlice[self.filterCol]):
             infilt = np.where(dataSlice[self.filterCol] == filtername)
-            mags[infilt] = self.lightcurves.interp(t[infilt], filtername, lc_indx=slicePoint['file_indx'])
+            mags[infilt] = self.lightcurves.interp(t[infilt], filtername,
+                                                   lc_indx=slicePoint['file_indx'])
             # Apply dust extinction on the light curve
             A_x = self.Ax1[filtername] * slicePoint['ebv']
-            mags[infilt] -= A_x
+            mags[infilt] += A_x
 
             distmod = 5*np.log10(slicePoint['distance']*1e6) - 5.0
             mags[infilt] += distmod
 
-        result['multi_detect'] = self._multi_detect(dataSlice, slicePoint, mags, t)
-        result['multi_color_detect'] = self._multi_color_detect(dataSlice, slicePoint, mags, t)
+        result['multi_detect'] = self._multi_detect(dataSlice, mags, t)
```

Comment:
May I make a suggestion? You don't have to do this but I notice that 'around peak' is similar for most of these detection criteria. Does it make sense to pick out the 'around_peak' indexes first, then pass that subset to each of these selection criteria, instead of the entire dataslice / entire mags / entire time?

Reply:
This is indeed a smart thing to do, thank you for the suggestion! I modified the code to reduce the number of variables passed.